apollo-federation/src/sources/connect/json_selection/parser.rs

```diff
@@ -84,22 +107,50 @@ impl ExternalVarPaths for JSONSelection {
     }
 }
 
-// NamedSelection       ::= NamedPathSelection | NamedFieldSelection | NamedQuotedSelection | NamedGroupSelection
+// NamedSelection       ::= NamedPathSelection | NamedFieldSelection | NamedGroupSelection
 // NamedPathSelection   ::= Alias PathSelection
-// NamedFieldSelection  ::= Alias? Identifier SubSelection?
-// NamedQuotedSelection ::= Alias LitString SubSelection?
+// NamedFieldSelection  ::= Alias? Key SubSelection?
 // NamedGroupSelection  ::= Alias SubSelection
 
 #[derive(Debug, PartialEq, Eq, Clone)]
 pub enum NamedSelection {
-    Field(Option<Alias>, String, Option<SubSelection>),
-    Quoted(Alias, String, Option<SubSelection>),
+    Field(Option<Alias>, WithRange<Key>, Option<SubSelection>),
     Path(Alias, PathSelection),
     Group(Alias, SubSelection),
 }
 
+// Like PathSelection, NamedSelection is an AST structure that takes its range
+// entirely from its children, so NamedSelection itself does not need to provide
+// separate storage for its own range, and therefore does not need to be wrapped
+// as WithRange<NamedSelection>, but merely needs to implement the Ranged trait.
+impl Ranged<NamedSelection> for NamedSelection {
+    fn node(&self) -> &NamedSelection {
+        self
+    }
```

Comment:
You could even add a trait bound to the trait itself to catch any instances where deref isn't implemented. Something like:

```rust
pub trait Ranged<T>: Deref<Target = T> {
  ...
}
```

Reply:
I'm implementing `Ranged<T>` for some types `T` that do not implement `Deref`, whose `node` methods return `self` (like `PathSelection` and `NamedSelection`), so I don't think that trait bound will work.

Something I could do, instead of requiring this `node` method, is to rely on the `AsRef` trait implemented for `WithRange<T>`, and just handle the non-`WithRange` implementations of `Ranged<T>` as usual (no need to call `.as_ref()`. There's already some ambiguity about when I use `wrapped.node()` vs. `wrapped.as_ref()`, so I suspect we can do without `node`.

Reply:
Successfully removed `Ranged<T>::node()`: https://github.com/apollographql/router/pull/5987/commits/4005495847798865a6544f62f4dc7a0898874ec7